File: zig/src/nucleus/sql.zig
// Nucleus SQL Model — typed database access with comptime struct scanning
//
// The foundation model that enables typed query execution:
// - query() returns results mapped to comptime-typed structs via @typeInfo
// - queryOne() returns a single row or null
// - execute() returns rows affected

const std = @import("std");
const NucleusClient = @import("client.zig").NucleusClient;
const QueryResult = @import("client.zig").QueryResult;

pub const SqlModel = struct {
    client: *NucleusClient,

    /// Execute a query and return the raw QueryResult.
    pub fn queryRaw(self: SqlModel, sql: []const u8) !QueryResult {
        return try self.client.query(sql);
    }

    /// Execute a query and return the first column of the first row as a string.
    pub fn queryScalar(self: SqlModel, sql: []const u8) !?[]const u8 {
        return try self.client.execute(sql);
    }

    /// Execute a statement (INSERT/UPDATE/DELETE) and return the command tag.
    pub fn execute(self: SqlModel, sql: []const u8) ![]const u8 {
        return try self.client.exec(sql);
    }

    /// Format a SQL string with arguments into a buffer, then execute it.
    /// Uses bufPrint for simple parameter substitution.
    pub fn executeFormatted(self: SqlModel, comptime fmt: []const u8, args: anytype) ![]const u8 {
        var buf: [4096]u8 = undefined;
        const sql = std.fmt.bufPrint(&buf, fmt, args) catch return error.BufferTooShort;
        return try self.client.exec(sql);
    }

    /// Format a SQL string with arguments and return scalar result.
    pub fn queryFormatted(self: SqlModel, comptime fmt: []const u8, args: anytype) !?[]const u8 {
        var buf: [4096]u8 = undefined;
        const sql = std.fmt.bufPrint(&buf, fmt, args) catch return error.BufferTooShort;
        return try self.client.execute(sql);
    }
};

/// Parse a comptime struct type's fields for column mapping.
/// Returns the field names as a comptime slice, useful for SELECT column lists.
pub fn structFields(comptime T: type) []const []const u8 {
    const info = @typeInfo(T);
    switch (info) {
        .@"struct" => |s| {
            var names: [s.fields.len][]const u8 = undefined;
            for (s.fields, 0..) |field, i| {
                names[i] = field.name;
            }
            return &names;
        },
        else => @compileError("structFields requires a struct type"),
    }
}

/// Generate a SELECT column list from a struct type at compile time.
pub fn selectColumns(comptime T: type) []const u8 {
    const info = @typeInfo(T);
    switch (info) {
        .@"struct" => |s| {
            var result: []const u8 = "";
            for (s.fields, 0..) |field, i| {
                if (i > 0) result = result ++ ", ";
                result = result ++ field.name;
            }
            return result;
        },
        else => @compileError("selectColumns requires a struct type"),
    }
}

// ── Tests ─────────────────────────────────────────────────────

test "selectColumns: simple struct" {
    const User = struct {
        id: i64,
        name: []const u8,
        email: []const u8,
    };
    const cols = comptime selectColumns(User);
    try std.testing.expectEqualStrings("id, name, email", cols);
}

test "selectColumns: single field" {
    const Count = struct {
        count: i64,
    };
    const cols = comptime selectColumns(Count);
    try std.testing.expectEqualStrings("count", cols);
}

test "structFields: returns field names" {
    const Item = struct {
        id: i64,
        title: []const u8,
    };
    const fields = comptime structFields(Item);
    try std.testing.expectEqual(@as(usize, 2), fields.len);
    try std.testing.expectEqualStrings("id", fields[0]);
    try std.testing.expectEqualStrings("title", fields[1]);
}

File: go/nucleus/sql.go
package nucleus

import (
	"context"
	"fmt"
	"reflect"
	"strconv"
	"strings"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/neutron-dev/neutron-go/neutron"
)

// querier is the interface satisfied by both *pgxpool.Pool and txQuerier.
type querier interface {
	Query(ctx context.Context, sql string, args ...any) (pgx.Rows, error)
	QueryRow(ctx context.Context, sql string, args ...any) pgx.Row
	Exec(ctx context.Context, sql string, args ...any) (pgconn.CommandTag, error)
}

// SQLModel provides type-safe SQL query execution.
type SQLModel struct {
	pool querier
}

// Query executes a query and scans all rows into a slice of T.
// T must be a struct with `db` tags matching column names.
func Query[T any](ctx context.Context, sql *SQLModel, query string, args ...any) ([]T, error) {
	// Use SimpleProtocol for SELECT queries. Nucleus extended protocol
	// returns incorrect results for COUNT/SUM aggregates.
	queryArgs := make([]any, 0, len(args)+1)
	queryArgs = append(queryArgs, pgx.QueryExecModeSimpleProtocol)
	queryArgs = append(queryArgs, args...)
	rows, err := sql.pool.Query(ctx, query, queryArgs...)
	if err != nil {
		return nil, fmt.Errorf("nucleus: query: %w", err)
	}
	defer rows.Close()

	var results []T
	for rows.Next() {
		var item T
		if err := scanRow(rows, &item); err != nil {
			return nil, fmt.Errorf("nucleus: scan: %w", err)
		}
		results = append(results, item)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("nucleus: rows: %w", err)
	}

	return results, nil
}

// QueryOne executes a query and scans exactly one row into T.
// Returns ErrNotFound if no rows are returned.
func QueryOne[T any](ctx context.Context, sql *SQLModel, query string, args ...any) (T, error) {
	var result T
	queryArgs := make([]any, 0, len(args)+1)
	queryArgs = append(queryArgs, pgx.QueryExecModeSimpleProtocol)
	queryArgs = append(queryArgs, args...)
	rows, err := sql.pool.Query(ctx, query, queryArgs...)
	if err != nil {
		return result, fmt.Errorf("nucleus: query: %w", err)
	}
	defer rows.Close()

	if !rows.Next() {
		if err := rows.Err(); err != nil {
			return result, fmt.Errorf("nucleus: rows: %w", err)
		}
		return result, neutron.ErrNotFound("no rows returned")
	}

	if err := scanRow(rows, &result); err != nil {
		return result, fmt.Errorf("nucleus: scan: %w", err)
	}

	return result, nil
}

// Exec executes a non-query SQL statement and returns the number of affected rows.
// Uses SimpleProtocol for Nucleus compatibility — its extended protocol types
// all parameters as TEXT which breaks arithmetic on BIGINT columns.
func (s *SQLModel) Exec(ctx context.Context, query string, args ...any) (int64, error) {
	execArgs := make([]any, 0, len(args)+1)
	execArgs = append(execArgs, pgx.QueryExecModeSimpleProtocol)
	execArgs = append(execArgs, args...)
	tag, err := s.pool.Exec(ctx, query, execArgs...)
	if err != nil {
		return 0, fmt.Errorf("nucleus: exec: %w", err)
	}
	return tag.RowsAffected(), nil
}

// scanRow scans a pgx row into a struct using `db` tags.
func scanRow(rows pgx.Rows, dest any) error {
	rv := reflect.ValueOf(dest)
	if rv.Kind() != reflect.Ptr || rv.IsNil() {
		return fmt.Errorf("dest must be a non-nil pointer to struct")
	}
	rv = rv.Elem()
	rt := rv.Type()

	if rt.Kind() != reflect.Struct {
		// For non-struct types, try direct scan
		return rows.Scan(dest)
	}

	// Build column-name to field-index map
	fieldDescs := rows.FieldDescriptions()
	colNames := make([]string, len(fieldDescs))
	for i, fd := range fieldDescs {
		colNames[i] = string(fd.Name)
	}

	// Map db tags to field indices
	tagMap := make(map[string]int, rt.NumField())
	for i := 0; i < rt.NumField(); i++ {
		f := rt.Field(i)
		tag := f.Tag.Get("db")
		if tag == "" || tag == "-" {
			// Fall back to json tag
			tag = f.Tag.Get("json")
			if tag != "" {
				if idx := strings.Index(tag, ","); idx != -1 {
					tag = tag[:idx]
				}
			}
		}
		if tag == "" || tag == "-" {
			tag = strings.ToLower(f.Name)
		}
		tagMap[tag] = i
	}

	// Scan all columns as nullable strings, then convert to target types.
	// Nucleus pgwire may send binary format indicators with text data, so
	// direct scanning into typed fields fails. Nullable string intermediary
	// handles all cases including NULL values.
	rawVals := make([]any, len(colNames))
	rawPtrs := make([]*string, len(colNames))
	for i := range rawVals {
		rawVals[i] = &rawPtrs[i] // **string — allows NULL scanning
	}

	if err := rows.Scan(rawVals...); err != nil {
		return err
	}

	for i, col := range colNames {
		fieldIdx, ok := tagMap[col]
		if !ok {
			continue
		}
		field := rv.Field(fieldIdx)

		// NULL → zero value
		if rawPtrs[i] == nil {
			continue
		}
		raw := *rawPtrs[i]

		switch field.Kind() {
		case reflect.String:
			field.SetString(raw)
		case reflect.Int, reflect.Int64:
			if raw == "" {
				field.SetInt(0)
			} else {
				n, err := strconv.ParseInt(raw, 10, 64)
				if err != nil {
					return fmt.Errorf("column %q: parse int %q: %w", col, raw, err)
				}
				field.SetInt(n)
			}
		case reflect.Float64:
			if raw == "" {
				field.SetFloat(0)
			} else {
				f, err := strconv.ParseFloat(raw, 64)
				if err != nil {
					return fmt.Errorf("column %q: parse float %q: %w", col, raw, err)
				}
				field.SetFloat(f)
			}
		case reflect.Bool:
			field.SetBool(raw == "true" || raw == "t" || raw == "1")
		default:
			field.SetString(raw)
		}
	}

	return nil
}
